Solutii laborator 3 FII Practic

mircea98ro opened the pull request:
Am rezolvat problema "caesar"

mircea98ro committed:
Caesar

Am rezolvat problema "caesar"

mateimicu commented:
Mesajul Pull Requestului nu mai este relevant.

mateimicu commented:
Ar trebui să dai squash între commiturile care rezolva o problema. Există un concept de commit atomic, e acel commit care face un singur lucrur, în cazul de față un commit ar trebui sa rezolve o problema.Tu ai mai multe commituri care rezolva aceeași problema (unele adauga rezolvarea, altele aplica standardul `PEP8`).

mircea98ro commented:
Soo.. how do i squash commits?
(O sa ma apuc sa citesc normele PEP8 inainte sa mai rezolv probleme)

mateimicu commented:
Okay, asta a fost laborator dar ideea ar fi urmatoarea : 
Începi un rebase interactiv (`rebase -i  <commit_hash>` ), acolo o să îți apară toate commiturile de le ultimul până la cel dat ca parametru ( în ordine inversă ). De acolo poti face mai multe lucruri cu fiecare commit în parte : `pick`, `squash` , `edit` ... .Daca alegi squash atunci el o să iți adauge schimbarile de la commitu care are `squash` la precedentul ( le va considera unu ).

Uite un exemplu : 

``` git
pick fa532db commit1
pick 139dd45 commit2
squash 35caf1e commit3
pick f420d20 commit4

# Rebase acdc78d..f420d20 onto acdc78d
#
# Commands:
#  p, pick = use commit
#  r, reword = use commit, but edit the commit message
#  e, edit = use commit, but stop for amending
#  s, squash = use commit, but meld into previous commit
#  f, fixup = like "squash", but discard this commit's log message
#  x, exec = run command (the rest of the line) using shell
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
# Note that empty commits are commented out
```

Ceea ce e de observat sunt liniile : 

```
pick 139dd45 commit2
squash 35caf1e commit3
```

Daca continuam cu rebaseul el o sa faca `squash` intre commit3 si commit2 adica peste commit2 o sa puna si commit3, toate o sa fie un singur commit.

Linkuri ajutatoare: 
https://robots.thoughtbot.com/git-interactive-rebase-squash-amend-rewriting-history
https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History

mircea98ro commented:
Ah, imi aduc aminte. Ok, rebase. Got it! Mersi

mircea98ro commented:
@costingalan @micumatei  Acum e ok?

mateimicu commented:
Ești pe calea cea bună,  nu ar trebui să existe un commit care aplica standardul PEP8. Fiecare commit ar trebui să respecte standardul de la inceput. Ce ar trebui sa faci tu e sa modific commitul care rezolva fiecare problemă, îl faci să respecte standardul si renunti la ultimul commit.
Starea commiturilor ar trebui sa fie asta  : 

1910757  - Caesar 
72c49b7 - Paranteze

Commitul care implementeaza Caesar ar trebui sa respecte PEP8 din prima, acelas lucru este valabil si pentru Paranteze.

mircea98ro commented:
Ok, fac rebase imediat.

mircea98ro committed:
Paranteze

Am rezolvat problema "paranteze".

mircea98ro committed:
Unic

Am rezolvat problema "unic" si am mai si eliberat din memorie intre timp.
Beat that!

mateimicu commented:
Super :+1:

mircea98ro commented:
Deci inca mai e nevoie de acele Labels?

mateimicu commented on python/solutii/popovici_mircea/caesar.py:
```diff
@@ -0,0 +1,48 @@
+    """
+    key = ord(mesaj[0]) - ord('a')
+    for ch in mesaj:
+        if (ord(ch) >= ord('a') and ord(ch) <= ord('z')):
```
`dir(str)` arată ca avem niste metode `str.isalpha()`  și `str.islower()`.

mateimicu commented on python/solutii/popovici_mircea/paranteze.py:
```diff
@@ -0,0 +1,69 @@
+    sunt folosite corespunzător.
+    """
+    p = []
+    i = -1
```
Poate `i` nu e un nume sugestiv.

mateimicu commented on python/solutii/popovici_mircea/unic.py:
```diff
@@ -0,0 +1,39 @@
+        1 2 3 2 1 - 3
+        1 1 1 2 2 - 1
+    """
+    return sum(istoric) % sum(range(1, max(istoric) + 1))
```
E doar un caz particular.Nu știm ca numerele sunt de la `1 - n`, in input valid ar fi `istoric = [ 4, 6, 5, 4, 4]`.
Funcțiile `sum` și  `max` fac fiecare o parcurgere.Funcția `range` în python2.x genereaza o lista, aici folosesti memorie suplimentară.

mateimicu commented on python/solutii/popovici_mircea/paranteze.py:
```diff
@@ -0,0 +1,69 @@
+    Verifică dacă toate parantezele din expresie
+    sunt folosite corespunzător.
+    """
+    p = []
```
Poate `p` nu e un nume sugestiv.

mateimicu commented on python/solutii/popovici_mircea/paranteze.py:
```diff
@@ -0,0 +1,69 @@
+    - (][][)    nu este bine
+    - [)]()[(]  nu este bine
+"""
+# pylint: disable=unused-argument
```
Linia aceasta nu mai este necesară.

mateimicu commented on python/solutii/popovici_mircea/caesar.py:
```diff
@@ -0,0 +1,48 @@
+un mesaj prin înlocuirea fiecărei litere cu litera de pe poziția aflată
+la un n pași de ea în alfabet (unde este n este un număr întreg cunoscut
+"""
+# pylint: disable=unused-argument
```
Linia aceasta nu mai este necesară.

mateimicu commented on python/solutii/popovici_mircea/unic.py:
```diff
@@ -0,0 +1,39 @@
+    II. Găsește cercetătorul ce stă peste program după o singură
+    parcurgere a listei și fără a aloca memorie suplimentară.
+"""
+# pylint: disable=unused-argument
```
Linia aceasta nu mai este necesară.

mircea98ro committed:
ICAO

Am rezolvat cele 2 cerinte pentru "icao".

mateimicu commented on python/solutii/popovici_mircea/unic.py:
```diff
@@ -0,0 +1,40 @@
+        1 1 1 2 2 - 1
+    """
+    for item in istoric:
+        if istoric.count(item) % 2 != 0:
```
`cout` parcurge lista. Și în cazul de față folosești parcurgi de mai multe ori.

mateimicu commented on python/solutii/popovici_mircea/from_icao.py:
```diff
@@ -0,0 +1,44 @@
+    """Funcția va primi calea către fișierul ce conține mesajul brut și
+    va genera un fișier numit icao_intrare ce va conține mesajul inițial.
+    """
+    i_file = open(mesaj, 'r')
```
Nu tratezi cazul in care nu s-a putut deschide fisierul sau nu a fost gasit.

mateimicu commented on python/solutii/popovici_mircea/paranteze.py:
```diff
@@ -0,0 +1,55 @@
+        if item == '(' or item == '[':
+            deschise.append(keys[item])
+        else:
+            if deschise.__len__() == 0:
```
de ce folosesti metoda magică `__len__` când poti folosi funcția `len`. Exemplu `len(deschise)`.

mircea98ro replied:
Pentru ca nu stiu diferenta dintre metoda magica si functie si pentru ca prin `dir(list)` am gasit `__len__`

mateimicu replied:
https://docs.python.org/2/library/functions.html#len

mateimicu commented on python/solutii/popovici_mircea/paranteze.py:
```diff
@@ -0,0 +1,55 @@
+                return False
+            if deschise[deschise.__len__() - 1] + keys[item] != 0:
+                return False
+            deschise.__delitem__(deschise.__len__()-1)
```
Poți folosi metoda `pop` pe care o au listele, în loc să folosești metoda magică `__delitem__`. Iți recomand să dai un `help(list.pop)` pentru mai multe informații.

mateimicu commented on python/solutii/popovici_mircea/to_icao.py:
```diff
@@ -0,0 +1,47 @@
+    va genera un fișier numit mesaj.icao_intrare ce va conține
+    mesajul scris folosind alfabetul ICAO.
+    """
+    i_file = open('icao_intrare', 'r')
```
Nu tratezi cazul în care fișierul nu este găsit.
